cmd/webhook-server/admission_controller.go

```diff
@@ -0,0 +1,154 @@
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"net/http"
+)
+
+const (
+	jsonContentType = `application/json`
+)
+
+var (
+	universalDeserializer = serializer.NewCodecFactory(runtime.NewScheme()).UniversalDeserializer()
+)
+
+// patchOperation is an operation of a JSON patch, see https://tools.ietf.org/html/rfc6902 .
+type patchOperation struct {
+	Op    string      `json:"op"`
+	Path  string      `json:"path"`
+	Value interface{} `json:"value,omitempty"`
+}
+
+// admitFunc is a callback for admission controller logic. Given an AdmissionRequest, it returns the sequence of patch
+// operations to be applied in case of success, or the error that will be shown when the operation is rejected.
+type admitFunc func(*v1beta1.AdmissionRequest) ([]patchOperation, error)
+
+// isKubeNamespace checks if the given namespace is a Kubernetes-owned namespace.
+func isKubeNamespace(ns string) bool {
+	return ns == metav1.NamespacePublic || ns == metav1.NamespaceSystem
+}
+
+// doServeAdmitFunc parses the HTTP request for an admission controller webhook, and -- in case of a well-formed
+// request -- delegates the admission control logic to the given admitFunc. The response body is then returned as raw
+// bytes.
+func doServeAdmitFunc(w http.ResponseWriter, r *http.Request, admit admitFunc) ([]byte, error) {
+	// Step 1: Request validation. Only handle POST requests with a body and json content type.
+
+	if r.Method != http.MethodPost {
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		return nil, fmt.Errorf("invalid method %s, only POST requests are allowed", r.Method)
+	}
+
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return nil, fmt.Errorf("could not read request body: %v", err)
+	}
+
+	if contentType := r.Header.Get("Content-Type"); contentType != jsonContentType {
+		w.WriteHeader(http.StatusBadRequest)
+		return nil, fmt.Errorf("unsupported content type %s, only %s is supported", contentType, jsonContentType)
+	}
+
+	// Step 2: Parse the AdmissionReview request.
+
+	var admissionReviewReq v1beta1.AdmissionReview
+
+	if _, _, err := universalDeserializer.Decode(body, nil, &admissionReviewReq); err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return nil, fmt.Errorf("could not deserialize request: %v", err)
+	} else if admissionReviewReq.Request == nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return nil, errors.New("malformed admission review: request is nil")
+	}
+
+	// Step 3: Construct the AdmissionReview response.
+
+	admissionReviewResponse := v1beta1.AdmissionReview{
+		Response: &v1beta1.AdmissionResponse{
+			UID: admissionReviewReq.Request.UID,
+		},
+	}
+
+	var patchOps []patchOperation
+	// Apply the admit() function only for non-Kubernetes namespaces. For objects in Kubernetes namespaces, return
+	// an empty set of patch operations.
+	if !isKubeNamespace(admissionReviewReq.Request.Namespace) {
+		patchOps, err = admit(admissionReviewReq.Request)
+	}
+
+	if err != nil {
```

Comment:
The err should always be checked immediately after the line of code where it was returned, otherwise you might be processing a different error. For example, if `isKubeNamespace()` returns true, this `err` wouldn't correspond to that call.

Reply:
Fixed